### ISSUE TITLE
multi-gpu training is slow in lightning

### ISSUE THREAD

**Issue by mutasem-mattar** [PyTorchLightning/pytorch-lightning#3329]
## 🐛 Bug
I have migrated my code from pytorch to lightning. However, I noticed the iteration time is almost double in lightning. I have tried with one server and two servers (each have 4 V100 GPUs). It is always slower in lightning. I am using ddp distributed backend. However, I have noticed the same with horovod as well. I passed number --gpus  for my script in the argument so I can set num_workers > 0.

### Expected behavior

I  expected that lightning is just a wrapper over pytorch. I dont get why I am having reduction in training speed.

### Environment

 - PyTorch: 1.6.0
- PyTorch-Lightning: 0.9.0
 - How you installed PyTorch: pip
 - GPU models and configuration: (2 nodes with 4 V100 GPUs each).

**Comment by Borda** [comment]
mind share code sample to reproduce, because we are running parity testing and we test that PL runs almost the same fast as vanilla PT

**Comment by vr25** [comment]
Hi,

I am using multi-gpu multi-node with "ddp" distributed backend and it is extremely slow.

I ran the following script on a single CPU, GPU, and multiple nodes + multiple GPUs, and the last one (multi-node multi-GPU) is extremely slow and I can't figure out why. I am following the code from [here](https://github.com/shotleft/how-to-python/blob/master/How%20it%20works%20-%20Bike%20Share%20Regression%20PyTorch%20Lightning.ipynb). You can find the code [here](https://github.com/vr25/misc_exps/blob/master/sample_03_gpu.py).

Thanks!

**Comment by jszym** [comment]
I'm having similar issues performance degradation when using DDP.

Using `--gpus 1` results in 2.85s/it , whereas when I run with the flag `--gpus 4` (no other changes) the performance balloons to >400s/it. The backend in the multi-gpu case is the default ddp_spawn, and the GPUs are V100 32GB.

I can't share the code at the moment, but would be happy to do so privately.

**Comment by Borda** [comment]
@jszym Can you pls ping someone from @PyTorchLightning/core-contributors

**Comment by zhutmost** [comment]
@Borda  Looks that I have met a similar problem. I am using ddp with Lightning on two GPUs (RTX TITAN + 2080Ti), but it is much slower than my another plain Pytorch implementation  (~ 2x time per batch).
Here are some phenomena I observed, and I hope they can help the Lightning developers.
1. The validation & test speed seems normal. (~ ResNet18 on imagenet, 0.26sec/batch with a batch size of 128 per card) But the training takes 0.67sec/batch, which is about 2x time cost compared to my Pytorch DataParallel implementation on the same GPU cards.
2. I profile the trainer by setting `profiler='simple'`, it logged as follows:
```log
Profiler Report

Action                  |  Mean duration (s)    |  Total time (s)
-----------------------------------------------------------------
on_fit_start            |  4.7545e-05           |  4.7545e-05
on_validation_start     |  0.00035234           |  0.00070467
on_validation_start     |  0.00035234           |  0.00070467
on_validation_epoch_start       |  9.8335e-06           |  1.9667e-05
on_validation_batch_start       |  3.2243e-05           |  0.006384
validation_step_end     |  8.8535e-05           |  0.01753
on_validation_batch_end |  5.3126e-05           |  0.010519
on_validation_epoch_end |  1.8655e-05           |  3.7311e-05
on_validation_end       |  0.18651              |  0.37301
on_train_start          |  0.00012992           |  0.00012992
on_epoch_start          |  6.8345e-05           |  6.8345e-05
on_train_epoch_start    |  8.707e-06            |  8.707e-06
get_train_batch         |  0.0010243            |  4.9257
on_batch_start          |  1.2278e-05           |  0.059046
on_train_batch_start    |  8.6223e-06           |  0.041464
training_step_end       |  1.3568e-05           |  0.06525
model_forward           |  0.06697              |  322.06
model_backward          |  0.14361              |  690.6
on_after_backward       |  1.6886e-05           |  0.081205
optimizer_step          |  0.64568              |  3105.1
on_batch_end            |  1.4737e-05           |  0.070869
on_train_batch_end      |  1.7089e-05           |  0.082181
on_epoch_end            |  1.0795e-05           |  1.0795e-05
on_train_epoch_end      |  9.989e-06            |  9.989e-06
on_train_end            |  4.6208e-05           |  4.6208e-05
```

As a baseline to be compared with, my Pytorch implementation without Lightning takes 0.26 and 0.33 sec/batch for validation and training, respectively.

My Lightning implementation can be found [here](https://github.com/zhutmost/neuralzip) and the Pytorch one can be downloaded [here](https://github.com/zhutmost/lsq-net)

**Comment by zhutmost** [comment]
In the above logged profiler information, the model_backward and optimizer_step takes significantly long time. I am not sure whether it is correct.

**Comment by zhutmost** [comment]
Noted that I find another ddp bug. The DDP can only run on GPUs with a contiguous index sequence from zero.
In other words, the trainer can be set to `Trainer(.., gpus=[0,1,2,...])`. Meanwhile `Trainer(.., gpus=[1,2,...])` or `Trainer(.., gpus=[1])` will cause a runtime NCCL error.

So although I have three GPU cards (0: RTX Titan, 1 & 2: 2080Ti), I cannot run my program with DDP backend on the two 2080Ti cards, because their indexes are 1 and 2 and don't start from 0.

Looks that this bug should be irrelevant to this issue. (Maybe I can open another issue.

**Comment by awaelchli** [comment]
@zhutmost make sure to run the latest PL version which has the device idx fix. Otherwise a separate issue with repro script would be appreciated. 

Regarding the profiler, as far as I know optimizer step includes the closure for backward so the number there is not what it looks like.

**Comment by zhutmost** [comment]
Thanks for your reply :D
Here is part of my `conda list` information. I works on 1.0.8.
```
pytorch                   1.7.0           py3.8_cuda11.0.221_cudnn8.0.3_0    pytorch
pytorch-lightning         1.0.8              pyhd8ed1ab_0    conda-forge
pytorch-lightning-bolts   0.2.5rc1                 pypi_0    pypi
```

I know little about the internal mechanism of Lightning, and I have no idea how to accelerate my training process.

**Comment by zhutmost** [comment]
If you need more information to debug this slowdown phenomenon, I can help dump more needed logs.

**Comment by zhutmost** [comment]
It seems that the option `terminate_on_nan=True` takes about 1/3 time. So I disable it and the program runs much faster.

**Comment by hanikh** [comment]
I have a similar problem. I ran my code on one GPU (RTX 3090) and two GPUs (both are RTX 3090) and in the former the speed of each iteration is almost two times more than the latter. 
does anyone find out what the reason is?

**Comment by williamFalcon** [comment]
can you share a colab to replicate using the boring model?

**Comment by williamFalcon** [comment]
@zhutmost did this resolve for you?

**Comment by zhutmost** [comment]
Oh I have installed a PCIe SSD on my server to replace my HDD RAID, and put the dataset to the new SSD. And looks that this issue solved (or nearly solved). and I dont see an obvious performance drop in an epoch now.

I guess it only happens on a server with low I/O speed?

**Comment by awaelchli** [comment]
@hanikh how did you measure that speed? if you are looking at the it/s in the progress bar, then note that the progress bar only shows progress for gpu 0. 

For demonstration, if we run these two scripts:
```bash
# peak it/s: 210, data size per gpu: 1874
python pl_examples/basic_examples/simple_image_classifier.py --gpus 1 --accelerator ddp

# peak it/s: 190 (per gpu), data size per gpu: 937
python pl_examples/basic_examples/simple_image_classifier.py --gpus 2 --accelerator ddp
```
iteration per sec (per gpu) is a bit lower for 2 gpus because of extra communication overhead, but the epoch still finishes about 2x faster because each gpu runs through half of the data in parallel.

There other factors that can influence the multi-gpu experience:
1. device i/o: multi-gpu means more disk i/o speed is required because more workers try to access the device at the same time. you may need to adjust the num_workers.
2. model size: if your model is too small, the gpu's will spend more time copying data and communicating than the actual computation of forward+backward. In this case multi-gpu can not benefit you. you could still try to increase batch size.

Also, don't fully trust the display in the progress bar. If the epoch is too short and/or the progress bar is idle on the first batch (due to some extra computations at the beginning of the epoch) it will not be very accurate.

You can set `profile=True` in Trainer to get some more detailed info at the end of training. 
If you can provide more details or a script I can run, I'm happy to help further.

**Comment by hanikh** [comment]
Dear Adrian Wälchli, thanks for your time and consideration. Here is a
colab of my work. I would be so grateful if you take a look at it:
https://colab.research.google.com/drive/1picpk0Is9sc7dEspIpTHJ1Unzu_r3wY5?usp=sharing

**Comment by hanikh** [comment]
I increased num_workers but I saw no improvement in speed for num_workers>2.
the model size is not small at all and I am training on a relatively big dataset so epochs are not short.
And in my case as I mentioned before, the speed (it/s) is almost halved with 2 GPUs (really different from a bit slower).

**Comment by awaelchli** [comment]
You could use the [profiler](https://pytorch-lightning.readthedocs.io/en/latest/advanced/profiler.html) to check which operations take the longest to execute and compare them to your single gpu run. 
Since you mention that your model is tiny, it is possible that DDP communication is adding a noticeable overhead. You could try [gradient accumulation](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#accumulate-grad-batches).

**Comment by zhaone** [comment]
Hi, I also met the issue when I refactor my code. My profiler on single GPU
```shell
Action                                  |  Mean duration (s)    |Num calls              |  Total time (s)       |  Percentage %         |
--------------------------------------------------------------------------------------------------------------------------------------
Total                                   |  -                    |_                      |  3041.8               |  100 %                |
--------------------------------------------------------------------------------------------------------------------------------------
run_training_epoch                      |  3027.0               |1                      |  3027.0               |  99.516               |
run_training_batch                      |  0.22559              |12752                  |  2876.7               |  94.573               |
optimizer_step_with_closure_0           |  0.22285              |12752                  |  2841.8               |  93.426               |
training_step_and_backward              |  0.22077              |12752                  |  2815.3               |  92.556               |
backward                                |  0.11783              |12752                  |  1502.6               |  49.399               |
model_forward                           |  0.10246              |12752                  |  1306.6               |  42.956               |
training_step                           |  0.10225              |12752                  |  1303.9               |  42.867               |
evaluation_step_and_end                 |  2.5673               |34                     |  87.289               |  2.8697               |
validation_step                         |  2.5672               |34                     |  87.283               |  2.8695               |
get_train_batch                         |  0.002099             |12753                  |  26.769               |  0.88004              |
fetch_next_train_batch                  |  0.0020834            |12753                  |  26.57                |  0.87351              |
on_train_epoch_end                      |  14.294               |1                      |  14.294               |  0.46992              |
```
On multiple GPUs
```shell
Action                                  |  Mean duration (s)    |Num calls              |  Total time (s)       |  Percentage %         |
--------------------------------------------------------------------------------------------------------------------------------------
Total                                   |  -                    |_                      |  1069.3               |  100 %                |
--------------------------------------------------------------------------------------------------------------------------------------
run_training_epoch                      |  1006.2               |1                      |  1006.2               |  94.101               |
run_training_batch                      |  0.59959              |1594                   |  955.75               |  89.385               |
optimizer_step_with_closure_0           |  0.59654              |1594                   |  950.89               |  88.93                |
training_step_and_backward              |  0.51149              |1594                   |  815.31               |  76.25                |
model_forward                           |  0.37577              |1594                   |  598.98               |  56.018               |
training_step                           |  0.37468              |1594                   |  597.25               |  55.857               |
backward                                |  0.1337               |1594                   |  213.12               |  19.931               |
evaluation_step_and_end                 |  3.6568               |6                      |  21.941               |  2.052                |
validation_step                         |  3.6565               |6                      |  21.939               |  2.0518               |
on_train_epoch_end                      |  15.9                 |1                      |  15.9                 |  1.487                |
get_train_batch                         |  0.0079656            |1595                   |  12.705               |  1.1882               |
fetch_next_train_batch                  |  0.0079413            |1595                   |  12.666               |  1.1846               |
```
My multi GPUs strategy is `ddp`. At first, I think that this issue is caused by ddp dataloader. However, according to the profiler, forward slow down my code heavily. On multiple GPUs, the forward time increases nearly 4 times ... 

How can I find the bottleneck? Any suggestion?

**Comment by JiyangZhang** [comment]
Any update on this?

**Comment by zhutmost** [comment]
> Any update on this?
My experience (maybe outdated) is to put the dataset on a PCIe SSD.

**Comment by zxk19981227** [comment]
Any updates on this project？

**Comment by pacowong** [comment]
Any updates?

**Comment by jiarenyf** [comment]
What the fuck, meeting this problem too ...

**Comment by RitaQian-westlake** [comment]
similar problem, the training time of one epoch in my GNN model increases from 4.5h (single gpu) to 12h+ (2 gpus). A reference is that the cost time of building graphs for one batch increases from 27s to 2min28s
